### PR TITLE
Improve annotation application performance

### PR DESCRIPTION
**Pull request by Viicos**
Also remove unnecessary schema copy.

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/_internal/_generate_schema.py [hunk 1]
```diff
@@ -2106,7 +2106,7 @@ def _apply_single_annotation(self, schema: core_schema.CoreSchema, metadata: Any
                     return self.defs.definitions[new_ref]
                 schema['ref'] = new_ref  # type: ignore
 
-        maybe_updated_schema = _known_annotated_metadata.apply_known_metadata(metadata, schema.copy())
+        maybe_updated_schema = _known_annotated_metadata.apply_known_metadata(metadata, schema)
```
Comment (Viicos) [inline pydantic/_internal/_generate_schema.py:2109]: `apply_known_metadata` already copies `schema`

File: pydantic/_internal/_generate_schema.py [hunk 2]
```diff
@@ -2134,14 +2134,15 @@ def _get_wrapped_inner_schema(
...
         def new_handler(source: Any) -> core_schema.CoreSchema:
-            schema = metadata_get_schema(source, get_inner_schema)
-            schema = self._apply_single_annotation(schema, annotation)
-            schema = self._apply_single_annotation_json_schema(schema, annotation)
+            if annotation_get_schema is not None:
+                schema = annotation_get_schema(source, get_inner_schema)
+            else:
+                schema = get_inner_schema(source)
+                schema = self._apply_single_annotation(schema, annotation)
+                schema = self._apply_single_annotation_json_schema(schema, annotation)
```
Comment (Viicos) [inline pydantic/_internal/_generate_schema.py:2145]: This way, we don't go through `_appy_single_annotation*` for annotations with a custom `__get_pydantic_core_schema__` as it is unnecessary.

**Comment by Viicos** [comment]
This mostly improves performance for `test_field_validators_serializers` (which is ignored in codspeed), or for any similar use case where annotations with a custom `__get_pydantic_core_schema__` method is used.

**Review by sydney-runkle — APPROVED** [review]
Great find, thanks for making reviewing the perf diff easy!

**Comment by sydney-runkle** [comment]
Just as a side note, I think there's still a lot to do with https://github.com/pydantic/pydantic/issues/10036, I'm happy to take a deeper look at this for v2.11 :)

**Pull request merged**